src/Interpreters/CollectJoinOnKeysVisitor.cpp

```diff
@@ -80,57 +100,36 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
         ASTPtr right = func.arguments->children.at(1);
         auto table_numbers = getTableNumbers(left, right, data);
 
-        if (table_numbers.first != table_numbers.second && table_numbers.first > 0 && table_numbers.second > 0)
-            data.new_on_expression_valid = true;
-
-        /**
-          * if this is an inner join and the expression related to less than 2 tables, then move it to WHERE
-          */
-        if (data.kind == ASTTableJoin::Kind::Inner
-            && (table_numbers.first == table_numbers.second || table_numbers.first == 0 || table_numbers.second == 0))
+        if (canMoveToWhere(table_numbers, data.kind))
         {
-            if (!data.new_where_conditions)
-                data.new_where_conditions = ast->clone();
-            else
-                data.new_where_conditions = makeASTFunction("and", data.new_where_conditions, ast->clone());
+            addAndTerm(data.new_where_conditions, ast);
         }
         else
         {
+            if (data.kind == ASTTableJoin::Kind::Inner)
```

Comment:
If this is not an Inner join, this ast will not added to data.new_on_expression, I'm not sure whether`explain syntax select ...`  will miss this ast or not.

Reply:
But in this case we won't change `table_join.on_expression` at all https://github.com/ClickHouse/ClickHouse/pull/19685/files#diff-03665017349505c5529e5d0eeeeae80081e064267d6035b94cd428baf11856ceR432